libbeat/esleg/eslegclient/connection.go

```diff
@@ -311,16 +314,113 @@ func (conn *Connection) Close() error {
+}
+
+// ProxyDialer is a dialer that can be registered to golang.org/x/net/proxy
+func (settings *httpClientProxySettings) ProxyDialer(_ *url.URL, forward proxy.Dialer) (proxy.Dialer, error) {
+	return transport.DialerFunc(func(_, address string) (net.Conn, error) {
+
+		// Headers given to the CONNECT request
+		hdr := settings.Headers.Headers()
+		if settings.URL.User != nil {
+			username := settings.URL.User.Username()
+			password, _ := settings.URL.User.Password()
+			if len(hdr) == 0 {
+				hdr = http.Header{}
+			}
+			hdr.Add("Proxy-Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(username+":"+password)))
+		}
+
+		req := &http.Request{
+			Method: "CONNECT",
+			URL:    &url.URL{Opaque: address},
+			Host:   address,
+			Header: hdr,
+		}
+
+		// Dial the proxy host
+		c, err := forward.Dial("tcp", settings.URL.Host)
+		if err != nil {
+			return nil, err
+		}
+
+		// Write the CONNECT request
+		err = req.Write(c)
+		if err != nil {
+			c.Close()
+			return nil, err
+		}
+
+		res, err := http.ReadResponse(bufio.NewReader(c), req)
+		if err != nil {
+			c.Close()
+			return nil, err
+		}
+		res.Body.Close()
+
+		if res.StatusCode != http.StatusOK {
+			c.Close()
+			return nil, fmt.Errorf("proxy server returned status code %d", res.StatusCode)
+		}
+
+		return c, nil
+	}), nil
+}
+
 func (conn *Connection) Test(d testing.Driver) {
+	testProxyDialer := func(d testing.Driver, forward transport.Dialer, settings *httpcommon.HTTPClientProxySettings) transport.Dialer {
+		switch scheme := settings.URL.Scheme; scheme {
+		case "http", "https":
+			proxy.RegisterDialerType(scheme, (&httpClientProxySettings{HTTPClientProxySettings: settings}).ProxyDialer)
+		}
+		dialer, err := transport.ProxyDialer(logp.L(), &transport.ProxyConfig{URL: settings.URL.String()}, forward)
+		d.Fatal("proxy", err)
+		return dialer
+	}
+
 	d.Run("elasticsearch: "+conn.URL, func(d testing.Driver) {
 		u, err := url.Parse(conn.URL)
 		d.Fatal("parse url", err)
 
 		address := u.Host
 
+		if proxyURL := conn.Transport.Proxy.URL; proxyURL != nil && !conn.Transport.Proxy.Disable {
+			d.Run("proxy", func(d testing.Driver) {
+				dialer := transport.TestNetDialer(d, conn.Transport.Timeout)
+
+				if proxyURL.Scheme == "https" {
+					tls, err := tlscommon.LoadTLSConfig(conn.Transport.TLS)
+					if err != nil {
+						d.Fatal("load tls config", err)
+					}
+					dialer = transport.TestTLSDialer(d, dialer, tls, conn.Transport.Timeout)
+				}
+
+				_, err := dialer.Dial("tcp", proxyURL.Host)
+				d.Fatal("dial up", err)
+			})
+		}
+
 		d.Run("connection", func(d testing.Driver) {
-			netDialer := transport.TestNetDialer(d, conn.Transport.Timeout)
-			_, err = netDialer.Dial("tcp", address)
+			var dialer transport.Dialer
+			if proxyURL := conn.Transport.Proxy.URL; proxyURL == nil || conn.Transport.Proxy.Disable {
+				dialer = transport.TestNetDialer(d, conn.Transport.Timeout)
+			} else {
+				dialer = transport.NetDialer(conn.Transport.Timeout)
```

Comment:
I'm trying to understand the logic here. Why is the choice between the `TestNetDialer` and `NetDialer` is based on the proxy URL/proxy being enabled?

Reply:
@belimawr That's because the TCP connectivity has already been examined by `TestNetDialer` if the peer is the proxy server. Thanks to your comment, I realize there's much room to improve readability in the Test function. I hope the new one is self-explanatory enough.